mozjs/js/rust/src/jsslotconversions.rs

```diff
@@ -0,0 +1,250 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Conversion of Rust values to and from JS Slots
+
+#![deny(missing_docs)]
+
+use conversions::{ConversionResult, ConversionBehavior, FromJSValConvertible,
+                  ToJSValConvertible};
+use jsapi::root::*;
+use jsval;
+use std::mem;
+
+/// A trait to record the number of slots required for the current type
+pub trait NumSlots {
+    /// Number of slots required to store Self in inheritance
+    const NUM_SLOTS: u32 = 1;
+
+    /// Number of slots required to store Self as pointer
+    const NUM_SLOTS_PTR: u32 = 1;
```

Comment:
Pointer is a bit miss leading. All I want to say is it's stored as ObjectValue(I kind of think this as a pointer). Have one extra field maybe not necessary. I have some related answer below. Pasted here: The NUM_SLOTS is more for inheritance. If a class is inherit from DOMQuad, it will have a offset of 4 slots before its own slots. But if a class just contain a DOMQuad, it will just place 1 slot for DOMQuad. NUM_SLOTS_PTR is for the object stored as ObjectValue.